### ISSUE TITLE
[FEATURE] Add util class for parsing ExtensionAction bytes

### ISSUE THREAD

**Issue by dbwiddis** [opensearch-project/opensearch-sdk-java#607]
### Is your feature request related to a problem?

[Don't Repeat Yourself](https://en.wikipedia.org/wiki/Don't_repeat_yourself)

While reviewing [JS #349](https://github.com/opensearch-project/job-scheduler/pull/349) I noticed a lot of repetition of some of the same code as in #588.

This is all general code for the common pattern we've chosen of appending a request class name as a string to its parameters in bytes.

### What solution would you like?

Copy these repeated features to a class in the same package as the ExtensionAction / ExtensionTransportAction / ExtensionActionRequest and related classes.  They might fit well in an existing class (ExtensionAction possibly) or in a new class (ExtensionActionUtil).  These include:
 - the UNIT_SEPARATOR byte constant
 - a method for finding the position of that separator in the byte array
 - methods for splitting the class string and param bytes
 - perhaps even a method to instantiate the class with the params

Basically, all the functions needed to do this:
https://github.com/opensearch-project/opensearch-sdk-java/blob/ebc684ae223e87f1e05712e086fc1dab911a3b5f/src/main/java/org/opensearch/sdk/handlers/ExtensionActionRequestHandler.java#L85-L96

### What alternatives have you considered?

Repeating ourselves.

### Do you have any additional context?

This somewhat doubles down on using Writeable and if we're considering switching to protobuf, then we may consider a different approach.

**Comment by Kuanysh-kst** [comment]
Hello @dbwiddis, I would like to take this issue

**Comment by dbwiddis** [comment]
Assigning it to you, @Kuanysh-kst but let's make sure we prioritize finishing up your currently open PRs!

**Comment by Kuanysh-kst** [comment]
@dbwiddis can I redo this code like this :
```
    public static ExtensionActionRequest createExtensionActionRequest(ActionRequest actionRequest) {
        try {
            return (ExtensionActionRequest) actionRequest;
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
        return null;
    }
```

**Comment by dbwiddis** [comment]
@Kuanysh-kst I'm missing what comment this is in a reply to, but generally speaking:
 - this is a whole method to just do a class casting, which can happen on one line.  I'm guessing there's more to it than this.
 - We don't print to stdout, if necessary we log something.
 - We would have to handle the null wherever we call this so it's not a good return value for utility functions unless well documented.

I suspect there is a bigger question here that I'm missing because this comment seems disconnected.

**Comment by Kuanysh-kst** [comment]
My problem is that I do not know how to properly handle the ActionRequest class since this class does not have variables to work with or an array of bytes that could be obtained through this class

**Comment by dbwiddis** [comment]
@Kuanysh-kst I think I got an email about a comment here but it may have been deleted.  Let me try to clarify all the places where I think a util is useful.  I'm going to use the Hello World `RestRemoteHelloAction` sample here.

Creating the request:  this is already handled with the constructor syntax, no util needed.
```java
RemoteExtensionActionRequest proxyActionRequest = new RemoteExtensionActionRequest(SampleAction.INSTANCE, sampleRequest);
```
This is passed to the `SDKTransportService` which also collects the response
```java
RemoteExtensionActionResponse response = sdkTransportService.sendRemoteExtensionActionRequest(request);
```
The `SDKTransportService` (1st code snippet above) serializes the request class name and its input bytes, suitable for reconstructing it later.    So this is a good util function, `RestExtensionActionRequest` to bytes to use as a param for `TransportActionRequestFromExtension`.

Upon receipt at the target extension, we have to deserialize, that is handled in `ExtensionActionRequestHandler` (second code snippet above).  This produces the desired final class from the name name (in this case, an instance of `SampleRequest`) but to make the variable generic enough, we have its superclass, `ActionRequest`.  This second snippet can be broken down into two functionalities:
 - Separating the byte stream into the class name and bytes
 - Instantiating the class

They aren't exactly in sequence in the above snippet but the pieces are there.  Separating the stream and bytes is
```java
int nullPos = indexOf(request.getRequestBytes(), RemoteExtensionActionRequest.UNIT_SEPARATOR);
String requestClassName = new String(Arrays.copyOfRange(request.getRequestBytes(), 0, nullPos + 1),
    StandardCharsets.UTF_8).stripTrailing();
StreamInput requestByteStream = StreamInput.wrap(
    Arrays.copyOfRange(request.getRequestBytes(), nullPos + 1, request.getRequestBytes().length)
);
```
(Note the `indexOf` can/should be its own util method).

And then instantiating the class is:
```java
Class<?> clazz = Class.forName(requestClassName);
Constructor<?> constructor = clazz.getConstructor(StreamInput.class);
actionRequest = (ActionRequest) constructor.newInstance(requestByteStream);
```

That's all we need, since the next thing we do is `client.execute()` which only requires that `ActionRequest` class methods, so the fact it's a subclass doesn't matter.

SO in this sequence there are 3 distinct util methods:
1. Request name+bytes = combined bytes (the first snippet)
2. An indexOf method that you can pass the separator to
3. (Part 1) A method to separate the bytes (using the indexof) into the class name and streaminput
3. (Part 2) Using the above two results to instantiate an `ActionRequest`

Ideally the two methods in part 3 would be separate methods, but since we are returning two bits of information that complicates things, so a single method returning an ActionRequest works.

**Comment by dbwiddis** [comment]
In addition to the above 3 methods which work for Extension-to-Extension, we have the Plugin-to-Extension version.

I noted that https://github.com/opensearch-project/job-scheduler/pull/349 contained many of the same methods:
 - it used `indexOf()` which is defined above
 - 
It has a new method to convert a `Writeable` to its byte stream:
```java
private static <T extends Writeable> byte[] convertParamsToBytes(T actionParams)
```
This isn't obvious but in the Extension case (Hello World) we implemented similar code in the constructor taking the sample request.   It could be made shorter using that `convertParamsToBytes` which does the same thing.
```java
    public RemoteExtensionActionRequest(ActionType<? extends ActionResponse> instance, ActionRequest request) {
        this.action = instance.getClass().getName();
        this.requestClass = request.getClass().getName();
        byte[] bytes = new byte[0];
        try (BytesStreamOutput out = new BytesStreamOutput()) {
            request.writeTo(out);
            bytes = BytesReference.toBytes(out.bytes());
        } catch (IOException e) {
            throw new IllegalStateException("Writing an OutputStream to memory should never result in an IOException.");
        }
        this.requestBytes = bytes;
    }
```
So this conversion of `Writeable` to its bytes is a fourth method that should be added to the util class.

**Comment by dbwiddis** [comment]
So to sum up the above 2 comments, the 4 methods needed are:
1. ActionRequest --> bytes from its params
2. ActionRequest name + paramBytes = bytes to send
3. Reverse of both 2 and 1 to reconstruct an ActionRequest (ideally we'd reverse 2 and reverse 1 separately but that's harder to do)
4. an indexOf util method that all of the above can use

**Comment by Kuanysh-kst** [comment]
@dbwiddis can I open a new PR again , there are a lot of problems in the old one

**Comment by dbwiddis** [comment]
> @dbwiddis can I open a new PR again , there are a lot of problems in the old one

You can, but it's generally a better idea to try to fix the old one.  Usually a rebase commit and resolving conflicts fixes a lot of the issues.

```bash
git checkout main
git pull upstream main
git checkout yourprbranch
git rebase main
```
This will show any conflicting files.  Resolve conflicts, then `git add` and `git rebase --continue`.

Once you're done, force push the result.

Then fix any compile errors and push the fix.

**Comment by minalsha** [comment]
@dbwiddis can we close this issue since the PR associated with this issue: https://github.com/opensearch-project/OpenSearch/pull/6969 is merged?
